Ladybird/MachPortServer.cpp

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) 2024, Andrew Kaster <akaster@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "MachPortServer.h"
+#include <AK/Debug.h>
+#include <LibWebView/Platform/ProcessStatisticsMach.h>
+#include <bsm/libbsm.h>
+
+namespace Ladybird {
+
+MachPortServer::MachPortServer()
+    : m_thread(Threading::Thread::construct([this]() -> intptr_t { thread_loop(); return 0; }, "MachPortServer"sv))
+    , m_server_port_name(ByteString::formatted("org.SerenityOS.Ladybird.helper.{}", getpid()))
+{
+    if (auto err = allocate_server_port(); err.is_error())
+        dbgln("Failed to allocate server port: {}", err.error());
+    else
+        start();
+}
+
+MachPortServer::~MachPortServer()
+{
+    stop();
+}
+
+void MachPortServer::start()
+{
+    m_thread->start();
+}
+
+void MachPortServer::stop()
+{
+    m_thread->detach();
+    m_should_stop.store(true, MemoryOrder::memory_order_release);
+}
+
+bool MachPortServer::is_initialized()
+{
+    return MACH_PORT_VALID(m_server_port.port());
+}
+
+ErrorOr<void> MachPortServer::allocate_server_port()
+{
+    m_server_port = TRY(Core::MachPort::create_with_right(Core::MachPort::PortRight::Receive));
+    TRY(m_server_port.insert_message_right(Core::MachPort::MessageRight::MakeSend));
+    TRY(m_server_port.register_with_bootstrap(m_server_port_name));
+
+    dbgln_if(MACH_PORT_DEBUG, "Success! we created and attached mach port {:x} to bootstrap server with name {}", m_server_port.port(), m_server_port_name);
+    return {};
+}
+
+void MachPortServer::thread_loop()
+{
+    while (!m_should_stop.load(MemoryOrder::memory_order_acquire)) {
+
+        WebView::ParentPortMessage message {};
+        message.header.msgh_size = sizeof(message);
+        message.header.msgh_local_port = m_server_port.port();
+
+        // Get the pid of the child from the audit trailer so we can associate the port w/it
+        mach_msg_options_t const options = MACH_RCV_MSG | MACH_RCV_TRAILER_TYPE(MACH_RCV_TRAILER_AUDIT) | MACH_RCV_TRAILER_ELEMENTS(MACH_RCV_TRAILER_AUDIT);
+
+        auto const ret = mach_msg(&message.header, options, 0, sizeof(message), m_server_port.port(), MACH_MSG_TIMEOUT_NONE, MACH_PORT_NULL);
+        if (ret != KERN_SUCCESS) {
+            dbgln("mach_msg failed: {}", mach_error_string(ret));
+            break;
+        }
+
+        auto pid = audit_token_to_pid(message.trailer.msgh_audit);
```

Comment:
FWIW, that's just `msgh_audit.val[5]`, not sure it's worth linking to an external library just for that.

Reply:
Apple's mach/message.h header says 
```
/*
 * The audit token is an opaque token which identifies
 * Mach tasks and senders of Mach messages as subjects
 * to the BSM audit system.  Only the appropriate BSM
 * library routines should be used to interpret the
 * contents of the audit token as the representation
 * of the subject identity within the token may change
 * over time.
 */
 ```
 So... I'm not sure if relying on that being field 5 is 'safe'? It's likely that by the time we care about following all of Apple's sandbox/private API rules we'll have a different solution for this anyway though 🤷